File: C++/Level.2/타겟넘버.cpp
#include <string>
#include <vector>

using namespace std;

int n;
int cnt;

void DFSUtil(int depth, int sum, vector<int> numbers, int target) {
    if (depth == n) {
        if (sum == target) {
            ++cnt; 
        }
        return; 
    }
    DFSUtil(depth + 1, sum + numbers[depth], numbers, target);
    DFSUtil(depth + 1, sum - numbers[depth], numbers, target);
}

void DFS(vector<int> numbers, int target) {
    n = numbers.size();
    cnt = 0;
    DFSUtil(0, 0, numbers, target);
}

int solution(vector<int> numbers, int target) {
    DFS(numbers, target);
    return cnt;
}

File: JavaScript/Level.2/타겟넘버.js
function DFS(numbers, target) {
    let cnt = 0;
    let n = numbers.length;
    function DFSUtil(depth, value) {
        if (depth === n) {
            if (value === target) {
                ++cnt;
            }
            return;
        }
        DFSUtil(depth + 1, value + numbers[depth]);
        DFSUtil(depth + 1, value - numbers[depth]);
    }
    DFSUtil(0, 0);
    return cnt;
}

function solution(numbers, target) {
    return DFS(numbers, target);
}
